### DISCUSSION TITLE
[Showcase] Externs in haxe language

### DISCUSSION THREAD

**Discussion by djaonourside** [uNetworking/uWebSockets.js#366] [Show and tell]
Hello! Just want to share externs of uWebSockets.js in haxe language https://github.com/djaonourside/uwebsockets-externs. Haxe can also compile to cpp and hashlink(haxe virtual machine) and it's possible to have the same code of using uWebSockets for node.js and native target. Native uWebSockets  externs are not ready(I am not a native developer and need help in this direction ). It would be nice you add section "other languages" in readme and notice about the extern lib. Feel free to close the issue after reading:) Thank you for your library.

**Comment by hst-m** [comment]
Someone just posted [benchmarks](https://github.com/djaonourside/uwebsockets-externs/issues/1) to your library and it shows less than 40k req/sec? Pretty slow. I'd recommend using Javascript with uWS.js or C with uWS instead of messing with this Haxe thing

**Comment by ghost** [comment]
I have no idea what haxe or externs is and it looks to be quite obscure. Unless haxe is number two most popular thing in the world there is no value in it.

**Comment by ghost** [comment]
My point is uWebSockets.js is because nodejs is incredibly popular, not because I have any specific love for nodejs (I kind of hate it).

**Comment by djaonourside** [comment]
@hst-m 
> Someone just posted [benchmarks](https://github.com/djaonourside/uwebsockets-externs/issues/1) to your library and it shows less than 40k req/sec? Pretty slow. I'd recommend using Javascript with uWS.js or C with uWS instead of messing with this Haxe thing

Haxe generated js in most cases is the same speed as I would write it by hand but I have strong typed and many useful features. The benchmark is just about uWS.js speed))) 

> I'd recommend using Javascript with uWS.js or C with uWS instead of messing with this Haxe thing

Thank you I don't think js or c are suitable for me. Keep that crooked languages with yourself if you like 
perversion of this kind ;)

**Comment by ghost** [comment]
> Haxe generated js in most cases is the same speed as I would write it by hand but I have strong typed and many useful features.

So it is like TypeScript or Flow? Okay, but I don't see the relevance here.

> Thank you I don't think js or c are suitable for me. Keep that crooked languages with yourself if you like
perversion of this kind ;)

Opinions aside, the reality is that C and JS are two of the most popular languages by far in this whole industry, so either you cry in a corner how twisted reality is or you go with the flow. I would like JS to burn in hell, but I kind of have to deal with it.

**Comment by djaonourside** [comment]
>, so either you cry in a corner how twisted reality is or you go with the flow. 

There is another way: for performance critical thing I can choose for example C but for business logic I prefer well-designed languages. Haxe allowes me to do that way)

@alexhultman I don't want any holy wars. I just want to thank the lib authors for the really well made thing.

**Comment by ghost** [comment]
Alright 👍

**Comment by djaonourside** [comment]
@hst-m As expected no difference in speed between haxe and pure js https://github.com/djaonourside/uwebsockets-externs/issues/1#issuecomment-698025130

**Comment by ghost** [comment]
Ah! I understand.

I looked in your src folder and okay I get it.

You have essentially made what in TypeScript is called definitions. Aka you have specified the static interfaces in these .hx files.

Okay, well yes then the performance should be identical because you are still using the same very V8 binding. There is no difference.

Problem is nobody is using Haxxe, so again it has kind of only niche value

**Comment by ghost** [comment]
I think these definitions are well written and nicely made, but again Haxe is entirely foreign and nobody is using it.

* https://githut.info/
* TIOBE index

So the reality of the situation is that it has no value for this project.

**Comment by djaonourside** [comment]
@alexhultman Unfortunately that's true haxe languages is not popular despite of its well-done design and features( I think even better than java in many cases). Unlike typescript it can be compiled not only into js but also into cpp(and many other platforms). I'll try to write cpp externs of your lib and share with you a performance test result if you wouldn't mind.

**Comment by ghost** [comment]
Languages aren't supposed to be perfect. All languages have strange legacy quirks. Look at Swedish - it is the most nonsensical Viking legacy of a language and has roots in basically all of Europe and even parts of Asia. It makes no sense in most cases, and it has more exceptions than rules. But many people speak it so it has value.

This is also why an ancient language like C is still the most important language of computing - because millions of people speak it and it runs in every single device ever created.

**Comment by djaonourside** [comment]
Ok I suppose it's a long time dispute. It's cool both of us has instrument/interface he likes. My message was just about  opportunity for interaction between different interfaces:)

**Comment by hst-m** [comment]
@djaonourside he needs to do better benchmarks, they are low because he is using `autocannon` instead of better performing `wrk`, I looked at his repo and ran the same tests, I get 160K req/sec with uWS.js with `wrk -c100 -d5 -t1 http://localhost:3000/hello` and only 48K with `autocannon -c 100 -d 5 localhost:3000/hello`. I show uWS.js being 15 times faster than express, your autocannon benchmark shows it being only 2 times faster. I did not run Haxe benchmarks but can assume they are the same as uWS.js, seems silly to me to compile to JS instead of just writing JS to start, its the easiest to use and learn so might as well use it https://githut.info/ Wow JS the <span>#</span>1 most actively used language, Haxe is <span>#</span>47 behind Fortran at <span>#</span>46, hmm now might be good time to switch to JS

**Comment by djaonourside** [comment]
@hst-m It's a separate discussion why big and easily maintainable projects with a reach business logic shouldn't be written in pure javascript or c. I agree that c lang is most suitable for low level infrastructural projects like this one. Popularity is not any argument for me. Vice versa all the unique and beautiful things are(were) created by the significant minority of people who are unknown for majority. All the low-grade bullshit is about majority and mainstream.  Do you know the name of antibiotic inventor? But probably you do know well hits of Lady Gaga or shit like that. After all I see no reason why serious people  should count copy pasted js github repos. Quantity will not become quality.

**Comment by hst-m** [comment]
if you are like me doing full stack apps JS is the best and only choice, vanilla JS for complex single page browser apps for mobile and desktop, JS for server coding, database coding (mongodb) and the large ecosystem with every library you could want, and the language has matured nicely with needed features added

**Comment by djaonourside** [comment]
>if you are like me doing full stack apps JS is the best and only choice,

Not only and not best. I have the full access to all js infrastructure with haxe(the externs I showed above is illustrating it). And also I have access to cpp, c#, java, jvm, python, lua infrastructure only with one well-designed strictly typed language with smart compiler with macro support(meta programming). Js is not well-designed with its raw dynamics and f*cking implicit type transformations that look like an endless source of shooting in the self foot. Js badly approaches complicated projects. #-sign for private fields - are you serious? And it's not legacy it's modern thing invented by loads of Dan Abramov's adepts. When I see first time the simplest js code ```function foo(o) {}``` I know nothing about "o" param. It's a shapless world I don't want to live in and support.

**Comment by hst-m** [comment]
> raw dynamics and implicit type transformations

the things you consider negative I consider positive, like everything else the quality of the code depends on who wrote it, 90% of JS is probably coded badly and looks like spaghetti mess. You mention Dan Abramov of react so this is probably why you have bad impression, react.js is bad adds unnecessary black box layer of code that only overcomplicates everything and slows it down. So you must not do any front-end browser coding otherwise you'd be using JS and not trying to compile to it, so you have different use case. Anyone doing browser coding will want to use JS in browser and being able to use on server is a plus

**Comment by ghost** [comment]
Whatever floats your boat(s)
